Improve ALL settings pages: sections, help text, validation

colemanw opened the pull request:
Overview
----------------------------------------
Bigly improves our settings pages and settings metadata.
**User-facing changes:**

- Settings pages are now broken into "sections" each with their own header, icon & optional description.
- Better validation of settings
- Better visibility of locked settings
- Improved help text
- Use number inputs for numeric settings

Screenshots
----------------------------------------
_All settings pages look a bit nicer, here's a few examples:_

| Before | After |
| ------ | ----- |
| <img  alt="image" src="https://github.com/user-attachments/assets/854d8fa0-4146-40ee-a58a-81dfa76531e7" /> |  <img  alt="image" src="https://github.com/user-attachments/assets/0baab447-163c-4b99-b9cb-681d6d3d928c" /> |
| <img width="1130" height="1830" alt="image" src="https://github.com/user-attachments/assets/4042a22b-f9fc-4087-9761-a10edd831108" /> | <img width="1132" height="1826" alt="image" src="https://github.com/user-attachments/assets/79dcc983-0663-405b-a54d-1b487451e1bf" /> |
| <img width="1130" height="1786" alt="image" src="https://github.com/user-attachments/assets/7ae3bea5-231e-4818-b5e2-2d017467f0f4" /> | <img width="1134" height="1818" alt="image" src="https://github.com/user-attachments/assets/9793bc4a-9042-4375-820d-9cd4fcf49885" /> |


Technical Details
----------------------------------------
**Dev-facing changes:**


- There were confusingly 3 implementations of a generic settings page: `CRM_Admin_Form_Generic`, `CRM_Admin_Form_Setting` and `CRM_Admin_Form_Preferences`.  The cleanest one, and the one recommended for use by extensions was `CRM_Admin_Form_Generic` so I have implemented all improvements in that class, and switched all our existing pages to use it.
- Changes to all 3 classes should generally be backward-compatible with extensions using them, unless they're doing something really weird.
- Generic form now auto-validates settings based on data type and `validate_callback` in setting metadata.
- Use metadata to place setting in a section of the page. This is safe for extensions to add now, as older versions of Civi will just ignore it.
- To add sections to a settings form use `hook_civicrm_preProcess`.
- For advanced-use tweaks to markup in a section, each is wrapped in `{crmRegion}`.
- A field template override can be defined in the setting metadata.
- Individual fields can still be overridden or added in the `buildForm` function with no need for a custom template.
- Extinguishes the dumpster fire that was the `help_text`. Help text was previously double-entered: once on the setting metadata (without it, the form wouldn't know there *was* any help to show) and then again in a `.hlp` file for the actual contents, even though they were identical to what was in the metadata (or not, it was a mess).
- New standards for `help_text`: only plain text is supported. An array of strings will be marked up as paragraphs. Use `help_markup` instead of `help_text` if you absolutely must use markup, but if you just need to add a link to some docs, use `help_doc_url` (takes the same params as `{docURL}`).

colemanw committed:
Settings - Convert Misc form to use Generic, add sections

colemanw committed:
Settings - Add automatic form validation

Validate CRM_Admin_Form_Generic according to the validate_callback in settings metadata

colemanw committed:
SettingField - Get help from metadata

Help text was previously double-entered: once on the setting metadata (without it, the form wouldn't know there was any help to show) and then again in a .hlp file for the actual contents, even though they were identical to what was in the metadata (or not, it was a mess).

Now the setting metadata is the only place to store `help_text` and it will be escaped as plain text unless provided as `help_markup`. An array of values for `help_text` is also accepted and will be formatted as paragraphs.

In cases where the text generation needs some logic, it can be implemented in a `metadata_callback`.

Followup commits will cleanup the help in setting metadata files and delete the redundant `.hlp` files.

colemanw committed:
SettingField - Support doc_url in setting metadata

Improves the docURL function, allows an arbitrary url to be supplied, and formats prettier output including an icon.

colemanw committed:
Settings - Cleanup help text (Core)

Also deleting unused .hlp file

colemanw committed:
Settings - Fix settingTrait to apply html_attributes to all element types

colemanw committed:
Settings - Move smartGroupCacheTimeout to Misc - Performance section

colemanw committed:
Settings - Convert search settings to use Generic form, add sections

colemanw committed:
Settings - Convert Component settings to Generic, add sections

colemanw committed:
Settings - Enable settings pages to be split into sections

Adds functionality to Generic settings form for splitting page into sections

Sections can be added using hook_civicrm_preProcess

colemanw committed:
Settings - Convert URL settings to Generic + sections

colemanw committed:
Settings - Improve the look of overridden settings

Instead of an annoying popup with no information about *which* fields are overridden, add an icon to each overridden field and a small description at the top of the form.

Disabling fields instead of freezing them looks nicer; Quickform frozen fields look like vintage ascii-art ^_^

colemanw committed:
Settings - Fix generic form to allow dynamic additions

It's possible that a setting field's `metadata_callback` will add or remove it from a form. In order to support that, the form must start with fully-resolved metadata for all settings, before applying filters.

Picks out and cleans up some functionality from `CRM_Admin_Form_SettingTrait` to make it more efficient, also renames the default section to 'default' instead of '', which gives it a more sensible css class.

colemanw committed:
Settings - Convert Debug settings to Generic + sections

colemanw committed:
Settings - Convert Date settings to Generic + sections

colemanw committed:
Settings - Convert Mail settings to Generic + sections

colemanw committed:
Settings - Convert CiviMail settings to Generic + sections

colemanw committed:
Settings - Convert SMTP settings to Generic

Incomplete refactor as the `mailing_backend` setting is pretty strange.

colemanw committed:
Settings - Convert Mapping settings to Generic

colemanw committed:
Settings - Convert Path settings to Generic

colemanw committed:
Settings - Convert UF settings to Generic

colemanw committed:
Settings - Convert Case settings to Generic

colemanw committed:
Settings - Convert Member settings to Generic

colemanw committed:
Settings - Convert Contribute settings to Generic

colemanw committed:
Settings - Convert Misc settings to Generic

Add option for settings to supply their own tpl, a nicer alternative to the previous practice of overriding the entire form's tpl.

colemanw committed:
Settings - Move more CKEditor config to extension, fix button visibility

colemanw committed:
Settings - Convert Address settings to Generic

Use custom tpls for the fields with tokens.
Maybe we'd want to add a token selector option to the generic form at some point?

colemanw committed:
Api3 - Don't validate setting default values

Reverting settings to their default values is not user input
so should not invoke validation callbacks.

Fixes failing `api_v3_SettingTest::testRevertAll` now that `CRM_Core_BAO_Setting::validatePath` has been added to some settings.

colemanw committed:
Settings - Convert Localization settings to Generic + sections

Enables overriding elements in the buildForm function - if already added, CRM_Admin_Form_Generic won't re-add it.

colemanw committed:
Settings - Deprecate redundant form classes

These form classes are no longer used in core.
Extension authors should also be encouraged to use CRM_Admin_Form_Generic.

ufundo commented:
> @colemanw The "Outbound Email" settings (/civicrm/admin/setting/smtp?reset=1) have some special situations with respect to mailing_backend (and to the encrypted smtpPassword), so I decided to try it out (r-run). 

I think could sidestep this with a dedicated page for the `mailing_backend` setting?

(It's a horrible setting... I think it would be ripe for replacing with a config entity. Could maybe be rolled into SiteEmailAddress - outbound credentials could vary depending on what email address you want to send from?)

colemanw committed:
Settings - Fix failure to save some settings

Before: Most of the settings at the top of the page (VERP Separator, etc) fail to save.

After: Fixed, added some fixmes about how weird this form is.

pfigel commented:
FWIW, it looks like a few core extension expect `description` to allow markup, e.g.:
https://github.com/greenpeace-cee/civicrm-core/blob/1b6e2172eb71c7546142c0218a7375cc64dfbb70/ext/authx/settings/authx.setting.php#L104-L105
and
https://github.com/greenpeace-cee/civicrm-core/blob/1b6e2172eb71c7546142c0218a7375cc64dfbb70/ext/iframe/settings/iframe.setting.php#L36

Not sure if it's worth doing the `_markup` thing for description as well, or just change the text/move it to `help_markup`?

colemanw commented:
Good point @pfigel: https://github.com/civicrm/civicrm-core/pull/33900
